specs/BrowserProcessExited.md

```diff
@@ -66,25 +78,46 @@ void AppWindow::CloseWebView(/* ... */) {
               ICoreWebView2Environment* sender,
               ICoreWebView2BrowserProcessExitedEventArgs* args) {
           COREWEBVIEW2_BROWSER_PROCESS_EXIT_KIND kind;
+          UINT32 pid;
           CHECK_FAILURE(args->get_BrowserProcessExitKind(&kind));
-
-          // Watch for graceful browser process exit. Let ProcessFailed event
-          // handler take care of failed browser process termination.
-          if (kind == COREWEBVIEW2_BROWSER_PROCESS_EXIT_KIND_NORMAL)
+          CHECK_FAILURE(args->get_BrowserProcessId(&pid));
+
+          // If a new WebView is created from this CoreWebView2Environment after
+          // the browser has exited but before our handler gets to run, a new
+          // browser process will be created and lock the user data folder
+          // again. Do not attempt to cleanup the user data folder in these
+          // cases. If this happens, the PID of the new browser process will be
+          // different to the PID of the older process, we check against the
+          // PID of the browser process to which our last CoreWebView2 attached.
+          if (pid == m_newestBrowserPid)
+          {
+            // Watch for graceful browser process exit. Let ProcessFailed event
+            // handler take care of failed browser process termination.
+            if (kind == COREWEBVIEW2_BROWSER_PROCESS_EXIT_KIND_NORMAL)
+            {
+              CHECK_FAILURE(
+                  m_webViewEnvironment->remove_BrowserProcessExited(m_browserExitedEventToken));
+              // Release the environment only after the handler is invoked.
+              // Otherwise, there will be no environment to raise the event when
+              // the collection of WebView2 Runtime processes exit.
+              m_webViewEnvironment = nullptr;
+              CleanupUserDataFolder();
+            }
+          }
+          else
           {
-            CHECK_FAILURE(
-                m_webViewEnvironment->remove_BrowserProcessExited(m_browserExitedEventToken));
-            // Release the environment only after the handler is invoked.
-            // Otherwise, there will be no environment to raise the event when
-            // the collection of WebView2 Runtime processes exit.
-            m_webViewEnvironment = nullptr;
-            CleanupUserDataFolder();
+            // The exiting process is not the last in use. Do not attempt cleanup
+            // as we might still have a webview open over the user data folder.
+            MessageBox(
```

Comment:
⟦boilerplate comment about blocking calls inside event handlers⟧

Note also that a real program probably shouldn't annoy the user by displaying messages about things the user doesn't care about.

Reply:
This sample was originally written before the broader discussion around blocking calls inside event handlers. I will update.

In this case, the user data folder cleanup is only ever attempted upon user action (menu item selection).

Reply:
Thanks please do.

Reply:
Updated to post as async task.